rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+  * For toolchains not covered by Bazel out of the box, as well as for
+    cross-build toolchains, a set of rules defining the toolchain for
+    Bazel must be developed.
+    * A .bzl file containing code describing the location of tools, API, features, etc.
+    * To make the defined toolchain known to Bazel, a BUILD file
+      containing rules associating the toolchain with build properties, thus enabling build-time toolchain resolution
+
+Specific tasks:
+* C/C++
+  * External libs. Most of these are already bazelized (minimal build on MacOS).
+    * jemalloc, libffi, libgmp, libomp, libpatch, libpq, libprocps, libre2, libsodium, openssl, zlib
+  * libsnark.
+    * A special case.  Coda uses a modified fork of libsnark; it
+      depends on a handful of C/C++ libs (libff, libfqfft, etc.).
+    * I have forked the dependencies on github and added Bazel support
+      to each (about 90% complete).  The ideal is to get the
+      maintainers to accept the Bazel-enabling patches.
+    * For Coda this approach may or may not be desirable.  The
+      alternative is to use the same standard approach as used by the
+      other external libs, namely an external BUILD file.
+  * Internal C/C++ code.  The Code codebase contains some OCaml code
+    that generates C/C++ code which it then compiles.  The task is to
+    convert this to C/C++ code and let Bazel handle any customizations
+    needed.
+  * CUDA.  Coda contains a little bit of CUDA code.  I assume Bazel
+    will easily handle this, but I do not have a GPU, and have never
+    worked with CUDA.
+  * Cross-compile toolchains.  I've got a minimal example working
+    using a crosstool-NG toolchain on a Mac. The task is to polish it
+    and port it to Linux and possibly Windows.
+* OCaml
+  * Learn enough OCaml/Dune to understand the OCaml/C++ interface source and build code
+  * Learn enough Dune to understand how it is used in the Coda codebase.
+  * Analyze existing Bazel libraries supporting OCaml
+    * https://github.com/jin/rules_ocaml
+    * https://github.com/ostera/rules_reason
+  * develop `rules_ocaml`
+    * Option 1: Integrate Bazel and Dune
+      * Analyze `rules_foreign_cc` to see how it handles integration of
+    configure and make.
+        * Prototype a few Minimal Working Examples of using Bazel together with Dune.
+        * Implement `rules_ocaml` using this integrational strategy
+    * Option 2: Bazel-only ruleset
+      * Analyze `rules_go` (Bazel rules for Golang) as a possible model for `rules_ocaml`
+      * Implement `rules_ocaml` using this integrational strategy
+  * Use the new `rules_ocaml` to implement bazel support for Coda codebase:
+    * External libs
+    * Internal OCaml code
+    * C/C++ interop code
+* Golang
+  * Coda uses the Golang implementation of libp2p. Google maintains an
+    official set of Bazel rules for Go ([rules_go](https://github.com/bazelbuild/rules_go), so this should be
+    straightforward.
+
```

Comment:
There is hidden work here that I want to make explicit. We will soon have a large number of crypto-related Rust dependencies in the build (example: https://github.com/o1-labs/marlin ). This will likely be almost as easy as Go to integrate. See https://github.com/google/cargo-raze and https://github.com/bazelbuild/rules_rust